dependency-specification.rst

```diff
@@ -0,0 +1,522 @@
+dependencies are only relevant in one platform, or only some versions are
+acceptable, so the language permits describing all these cases.
+
+The language defined is a compact line based format which is already in
+widespread use in pip requirements files, though we do not specify the command
+line option handling that those files permit. There is one caveat - the
+URL reference form, specified in PEP-440 [#pep440]_ is not actually
+implemented in pip, but since PEP-440 is accepted, we use that format rather
+than pip's current native format.
+
+Motivation
+==========
+
+Any specification in the Python packaging ecosystem that needs to consume
+lists of dependencies needs to build on an approved PEP for such, but
+PEP-426 [#pep426]_ is mostly aspirational - and there are already existing
+implementations of the dependency specification which we can instead adopt.
+The existing implementations are battle proven and user friendly, so adopting
+them is arguably much better than approving an aspirational, unconsumed, format.
+
+Specification
+=============
+
+Examples
+--------
+
+All features of the language shown with a name based lookup::
+
+    requests [security,tests] >= 2.8.1, == 2.8.* ; python_version < "2.7.10"
+
+A minimal URL based lookup::
+
+    pip @ https://github.com/pypa/pip/archive/1.3.1.zip#sha1=da9234ee9982d4bbb3c72346a6de940a148ea686
+
+Concepts
+--------
+
+A dependency specification always specifies a distribution name. It may
+include extras, which expand the dependencies of the named distribution to
+enable optional features. The version installed can be controlled using
+version limits, or giving the URL to a specific artifact to install. Finally
+the dependency can be made conditional using environment markers.
+
+Grammar
+-------
+
+We first cover the grammar briefly and then drill into the semantics of each
+section later.
+
+A distribution specification is written in ASCII text. We use a parsley
+[#parsley]_ grammar to provide a precise grammar. It is expected that the
+specification will be embedded into a larger system which offers framing such
+as comments, multiple line support via continuations, or other such features.
+
+The full grammar including annotations to build a useful parse tree is
+included at the end of the PEP.
+
+Versions may be specified according to the PEP-440 [#pep440]_ rules. (Note:
+URI is defined in std-66 [#std66]_::
+
+    version_cmp   = wsp* '<' | '<=' | '!=' | '==' | '>=' | '>' | '~=' | '==='
+    version       = wsp* ( letterOrDigit | '-' | '_' | '.' | '*' )+
+    version_one   = version_cmp version
+    version_many  = version_one (wsp* ',' version_one)*
+    versionspec   = ( '(' version_many ')' ) | version_many
+    urlspec       = '@' wsp* <URI_reference>
+
+Environment markers allow making a specification only take effect in some
+environments::
+
+    marker_op     = version_cmp | 'in' | 'not in'
+    python_str_c  = (wsp | letter | digit | '(' | ')' | '.' | '{' | '}' |
```

Comment:
This also needs expanded since it can't contain all the values we need just for version numbers, but probably it should be expanded to handle more than just version numbers too yea? Right now you can't do something like ``platform_version == '#3 on degas'`` which is in the Python documentation as an example for that. This is also restricted to only ASCII characters, is that reasonable or is there a reasonable value for one of the variables we have below that doesn't fall inside ascii?

Either way, this probably does need to support all the printable ascii characters at a minimum and might need to handle backslash escapes for quotes too since a lot of these strings could be completely freeform.